File: bfs_dfs/유기농배추_bfs.py
import sys
from collections import deque

input = sys.stdin.readline

T = int(input())

dr = [1, -1, 0, 0]
dc = [0, 0, 1, -1]


def bfs(s_row, s_col):
    queue = deque()
    queue.append((s_row, s_col))
    visited[s_row][s_col] = 1

    while queue:
        r, c = queue.popleft()

        for i in range(4):
            nr = r + dr[i]
            nc = c + dc[i]

            if nr < 0 or nc < 0 or nr >= len(board) or nc >= len(board[0]):
                continue

            if visited[nr][nc] == 1:
                continue

            if board[nr][nc] == 0:
                continue

            queue.append((nr, nc))
            visited[nr][nc] = 1


for _ in range(T):
    M, N, K = map(int, input().split())
    board = [[0] * M for _ in range(N)]
    visited = [[0] * M for _ in range(N)]

    for _ in range(K):
        c, r = map(int, input().split())
        board[r][c] = 1

    answer = 0
    for row in range(len(board)):
        for col in range(len(board[0])):
            if board[row][col] == 1 and visited[row][col] == 0:
                bfs(row, col)
                answer += 1

    print(answer)

File: bfs_dfs/유기농배추_bfs.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <tuple>
#define REP(i, n) for (int i=0; i < n; i++)
using namespace std;

int dr[4] = {1, -1, 0, 0};
int dc[4] = {0, 0, 1, -1};

void bfs(int r, int c, vector<vector<int>>& board, vector<vector<int>>& visited){
  queue<vector<int>> queue;
  queue.push({r, c});
  visited[r][c] = 1;

  while (!queue.empty()) {
    vector<int> value = queue.front();
    queue.pop();
    int row = value[0];
    int col = value[1];

    REP(i, 4) {
      int nr = row + dr[i];
      int nc = col + dc[i];

      if (nr < 0 || nc < 0 || nr >= board.size() || nc >= board[0].size()) {
        continue;
      }

      if (visited[nr][nc] == 1) continue;

      if (board[nr][nc] == 0) continue;

      queue.push({nr, nc});
      visited[nr][nc] = 1;

    }
  }
}

int main(){
  int T;
  cin >> T;
  REP(i, T) {
    int M, N, K;
    cin >> M >> N >> K;

    vector<vector<int>> board(N, vector<int>(M, 0));
    vector<vector<int>> visited(N, vector<int>(M, 0));

    REP(i, K) {
      int c, r;
      cin >> c >> r;
      board[r][c] = 1;
    }

    int answer = 0;
    REP(row, N) {
      REP(col, M) {
        if (board[row][col] == 1 && visited[row][col] == 0) {
          bfs(row, col, board, visited);
          answer += 1;
        }
      }
    }
    cout << answer << endl;
  }
}
